algebird-core/src/main/scala/com/twitter/algebird/RingBuf.scala

```diff
@@ -0,0 +1,105 @@
+package com.twitter.algebird
+
+/**
+ * Value class to hold references to indices internal to RingBuf.
+ */
+case class Idx(value: Int) extends AnyVal {
+  def +(i: Int): Idx = Idx(value + i)
+  def -(i: Int): Idx = Idx(value - i)
+
+  // Look/ itself up in the supplied vector.
+  def of[A](v: Vector[A]): A = v(value)
+
+  // updates the supplied vector with the supplied value.
+  def update[A](v: Vector[A], a: A): Vector[A] = v.updated(value, a)
+}
+
+object Idx {
+  def bounded(i: Long, bound: Int): Idx = Idx(((i + bound) % bound).toInt)
+}
+
+case class RingBuf[A](slots: Vector[A], index: Idx) { lhs =>
+  import RingBuf._
+  /**
+   * Slots are 0-indexed from the current `index`, looking backwards.
+   *
+   * `toIdx` converts an index into a vector position in `slots` for
+   * internal use.
+   *
+   * i=1, index=5, toIdx(i)=Idx(4)
+   * i=(slots.size - 1), index=5, toIdx(i)==Idx(6) (ie last item)
+   */
+  private[this] def toIdx(i: Long): Idx = rawIdx(i + index.value)
+
+  /**
+   * Use if you don't need the `index.value` offset.
+   */
+  private[this] def rawIdx(i: Long): Idx = Idx.bounded(i, slots.size)
+
+  private[this] def assertBounds(i: Long): Unit =
+    assert(i >= 0 && i < slots.size, s"$i is out of allowed bounds: [0 $slots.size)")
+
+  /**
+   * Returns a copy of the RingBuf with `a` added in to the proper
+   * externally-indexed slot.
+   */
+  def add(a: A, i: Long)(implicit ev: Monoid[A]): RingBuf[A] = {
+    assertBounds(i)
+    val idx = toIdx(i)
+    copy(slots = idx.update(slots, ev.plus(idx.of(slots), a)))
+  }
+
+  /**
+   * Look up the item in the vector by external index.
+   */
+  def apply(i: Long): A = {
+    assertBounds(i)
+    toIdx(i).of(slots)
+  }
+
+  /**
+   * Steps the RingBuf forward `i` steps, zero-ing out all new
+   * entries.
+   */
+  def step(i: Long)(implicit ev: Monoid[A]): RingBuf[A] = {
+    if (i <= 0) this
+    else {
+      val v = (1L to (i min slots.size)).foldLeft(slots) {
+        (acc, j) => toIdx(j).update(acc, ev.zero)
+      }
+      RingBuf(v, toIdx(i))
+    }
+  }
+
+  def +(rhs: RingBuf[A])(implicit ev: Monoid[A]): RingBuf[A] = {
```

Comment:
OH! Forgot to mention. This actually fixes a bug in the other implementation.

Before, we assumed that both sides of the addition had the same size backing vector. We were starting from `index + 1` and wrapping around.

If you add vectors of different sizes this breaks. I changed it to start from the head and walk backward, which will truncate the longer `RingBuf` instance but maintains the associative property.